spd/metrics/pgd_utils.py

```diff
@@ -0,0 +1,188 @@
+                )
+        mask_infos = make_mask_infos(
+            component_masks=_interpolate_component_mask(ci, component_sample_points),
+            weight_deltas_and_masks=weight_deltas_and_masks,  # we don't interpolate for the weight delta
+            routing_masks=routing_masks,
+        )
+        out = model(batch, mask_infos=mask_infos)
+        total_loss = calc_sum_recon_loss_lm(pred=out, target=target_out, loss_type=output_loss_type)
+        n_examples = out.shape.numel() if output_loss_type == "mse" else out.shape[:-1].numel()
+        return total_loss / n_examples
+
+    *batch_dims, C = next(iter(ci.values())).shape
+
+    def maybe_repeat_to_batch_shape(
+        component_sample_points: dict[str, Float[Tensor, "... C"]],
+        weight_delta_mask: dict[str, Float[Tensor, "..."]] | None,
+    ) -> tuple[dict[str, Float[Tensor, "... C"]], dict[str, Float[Tensor, "..."]] | None]:
+        return (
+            {k: v.expand(*batch_dims, -1) for k, v in component_sample_points.items()},
+            {k: v.expand(*batch_dims) for k, v in weight_delta_mask.items()}
+            if weight_delta_mask is not None
+            else None,
+        )
+
+    match pgd_config.mask_scope:
+        case "unique_per_datapoint":
+            ci_mask_shape = torch.Size(batch_dims) + (C,)
+            weight_delta_mask_shape = torch.Size(batch_dims)
+        case "shared_across_batch":
+            # replace all batch dims with 1 so we broadcast the same mask across all datapoints
+            ci_mask_shape = torch.Size(1 for _ in batch_dims) + (C,)
+            weight_delta_mask_shape = torch.Size(1 for _ in batch_dims)
+
+    adversarial_component_sample_points: dict[str, Float[Tensor, "... C"]] = {}
+    for layer in ci:
+        sample_point = _get_pgd_init_tensor(pgd_config.init, ci_mask_shape, batch.device)
+        sample_point.requires_grad_(True)
+        adversarial_component_sample_points[layer] = sample_point
+
+    adversarial_weight_delta_masks: dict[str, Float[Tensor, " ..."]] | None = None
+    if weight_deltas is not None:
+        adversarial_weight_delta_masks = {}
+        for layer in ci:
+            wd_init = _get_pgd_init_tensor(pgd_config.init, weight_delta_mask_shape, batch.device)
+            wd_init.requires_grad_(True)
+            adversarial_weight_delta_masks[layer] = wd_init
+
+    # PGD ascent
+    adv_vars = list(adversarial_component_sample_points.values())
+    if adversarial_weight_delta_masks is not None:
+        adv_vars.extend(list(adversarial_weight_delta_masks.values()))
+
+    for _ in range(int(pgd_config.n_steps)):
+        assert all(v.requires_grad for v in adv_vars)
+        assert all(v.grad is None for v in adv_vars)
+
+        with torch.enable_grad():
+            obj = objective_fn(
+                *maybe_repeat_to_batch_shape(
+                    adversarial_component_sample_points, adversarial_weight_delta_masks
```

Comment:
I'd be tempted to use "adv" instead of "adversarial" throughout this function. I think it would reduce the space in places like this.